src/simulation.hpp

```diff
@@ -1021,6 +1079,131 @@ template <typename problem_t> void AMRSimulation<problem_t>::ellipticSolveAllLev
+					 amrex::GeometryData const &geom, const amrex::Real &time, const amrex::BCRec *bcr, int bcomp,
+					 const int &orig_comp) const
+	{
+		amrex::ignore_unused(iv, dest, dcomp, numcomp, geom, time, bcr, bcomp, orig_comp);
+	}
+};
+
+template <typename problem_t> void AMRSimulation<problem_t>::kickParticlesAllLevels(const amrex::Real dt)
+{
+	// kick particles (do: vel[i] += 0.5 * dt * accel[i])
+
+	if (do_cic_particles != 0) {
+		// gravitational acceleration multifabs
+		amrex::Vector<amrex::MultiFab> accel(finest_level + 1);
+
+		// self-gravity in Quokka requires open boundary conditions,
+		// so we extrapolate the gravitational accelerations at physical boundaries
+		amrex::Vector<amrex::BCRec> accelBC(AMREX_SPACEDIM);
+		for (int j = 0; j < AMREX_SPACEDIM; ++j) {
+			for (int i = 0; i < AMREX_SPACEDIM; ++i) {
+				accelBC[j].setLo(i, amrex::BCType::foextrap);
+				accelBC[j].setHi(i, amrex::BCType::foextrap);
+			}
+		}
+
+		for (int lev = 0; lev <= finest_level; ++lev) {
+			// compute accelerations
+			accel[lev].define(boxArray(lev), DistributionMap(lev), AMREX_SPACEDIM, 1);
+			accel[lev].setVal(0.);
+			auto accel_arr = accel[lev].arrays();
+			const auto &phi_arr = phi[lev].const_arrays();
+			const auto dx_inv = geom[lev].InvCellSizeArray();
+			const amrex::IntVect ng(0);
+
+			// check for NaN
+			AMREX_ALWAYS_ASSERT(!phi[lev].contains_nan());
+
+			amrex::ParallelFor(accel[lev], ng, AMREX_SPACEDIM, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k, int n) {
+				// compute cell-centered acceleration -grad(phi)
+				if (n == 0) {
+					accel_arr[bx](i, j, k, n) = -0.5 * dx_inv[0] * (phi_arr[bx](i + 1, j, k) - phi_arr[bx](i - 1, j, k));
+				}
+				if (n == 1) {
+					accel_arr[bx](i, j, k, n) = -0.5 * dx_inv[1] * (phi_arr[bx](i, j + 1, k) - phi_arr[bx](i, j - 1, k));
+				}
+				if (n == 2) {
+					accel_arr[bx](i, j, k, n) = -0.5 * dx_inv[2] * (phi_arr[bx](i, j, k + 1) - phi_arr[bx](i, j, k - 1));
+				}
+			});
+			amrex::Gpu::streamSynchronizeAll();
+
+			// fill ghost cells for accel[lev]
+			amrex::GpuBndryFuncFab<setFunctorParticleAccel> boundaryFunctor(setFunctorParticleAccel{});
+			amrex::PhysBCFunct<amrex::GpuBndryFuncFab<setFunctorParticleAccel>> fineBoundaryFunctor(geom[lev], accelBC, boundaryFunctor);
+
+			if (lev == 0) {
+				accel[lev].FillBoundary(geom[lev].periodicity());
+				fineBoundaryFunctor(accel[lev], 0, accel[lev].nComp(), accel[lev].nGrowVect(), 0., 0);
+			} else {
+				amrex::PhysBCFunct<amrex::GpuBndryFuncFab<setFunctorParticleAccel>> coarseBoundaryFunctor(geom[lev - 1], accelBC,
+															  boundaryFunctor);
+				amrex::Vector<amrex::MultiFab *> fineData{&accel[lev]};
+				amrex::Vector<amrex::MultiFab *> coarseData{&accel[lev - 1]};
+				// N.B.: all multifabs are at the same time, so we ignore the time arguments
+				amrex::FillPatchTwoLevels(accel[lev], 0., coarseData, {0.}, fineData, {0.}, 0, 0, AMREX_SPACEDIM, geom[lev - 1], geom[lev],
```

Comment:
This can be simplified by instead using `amrex::InterpFromCoarseLevel`: https://github.com/AMReX-Codes/amrex/blob/022f97ea9ebbc28fca25f38f3eda26ce68b6230b/Src/AmrCore/AMReX_FillPatchUtil_I.H#L934.

Reply:
Done.